### ISSUE TITLE
HUGE performance issues with `ethers` (crypto, ArrayBuffer, ...)

### ISSUE THREAD

**Issue by mrousavy** [facebook/hermes#626]
## Bug Description

Using the [ethers.js library](https://docs.ethers.io/v4/cookbook-react.html#shims) is **extremely** slow on Hermes.

On my iPhone 11 Pro (very new and fast phone!) it takes: **half a minute** to create a new wallet:

```
 LOG     [16:49:19.253]	{?}	💰 Creating new Wallet...
 LOG     [16:49:52.812]	{?}	💰 New wallet created! Took 33559.440958321095ms, Phrase: shrug occur know victory derive pilot vivid heart time wise gentle angry
```

Whereas it takes "only" 14 seconds on JSC:

```
 LOG     [16:56:18.408]	{?}	💰 Creating new Wallet...
 LOG     [16:56:32.854]	{?}	💰 New wallet created! Took 14446.050416648388ms, Phrase: liquid museum speed design embark palm resemble hobby enough eager welcome sniff
```

And while JSC is twice as fast as Hermes, this is still extremely slow!

Reproduceable Code:

```ts
import 'react-native-get-random-values'
import '@ethersproject/shims'
import { ethers } from 'ethers'


  const now = performance.now()
  Logger.log('💰 Creating new Wallet...')
  const wallet = ethers.Wallet.createRandom()
  const end = performance.now()
  Logger.log(
    `💰 New wallet created! Took ${end - now}ms, Phrase: ${wallet.mnemonic.phrase}`,
  )
```

I believe it's one of the polyfills from `@ethersproject/shims` that's causing this huge slowdown. I'd love to find out how this performs without polyfills - that library needs native support for:

* `ArrayBuffer.isView`
* `atob` and `btoa`
* `nextTick`
* `Promise` 
* `String.prototype.normalize`
* `FileReader.prototype.readAsArrayBuffer`

As far as I know, `ArrayBuffer.isView` is supported by Hermes, `Promise` is polyfilled by React Native, and `String.prototype.normalize` is not used for `createWallet` (not sure about that)

- [x] I have disabled Hermes in Podfile, re-ran `pod install` and confirmed this bug does not occur with JSC

Hermes version: 0.9.0
React Native version (if any): 0.66.1
OS version (if any): iOS 15.0
Platform (most likely one of arm64-v8a, armeabi-v7a, x86, x86_64): arm64-v8a (I believe?)

**Comment by mrousavy** [comment]
I blindly tried to use @craftzdog's [`atob` and `btoa` polyfills](https://github.com/craftzdog/react-native-quick-base64) to see if the performance bottleneck arises from the Base64 operations, but that doesn't seem to have changed a thing.

I also added `console.log` statements in each of the polyfills from `@ethersproject/shims`, but none of them were called. The `createWallet()` function can't be _that_ slow (14s), I'll debug this further. Either way, Hermes is twice as slow as JSC in that case, so I'm leaving this bug report open

**Comment by neildhar** [comment]
This is similar to https://github.com/facebook/hermes/issues/569. Hermes is not going to be well suited to computationally intensive operations since it does not have a JIT. On iOS on a physical device, JSC doesn't have a JIT either, so it's going to be pretty slow as well. You'll actually likely see much better performance on an iOS emulator with JSC, because JSC is able to JIT there.

Your best option in such cases to get acceptable performance is probably to implement those computations in C++ and invoke them through JSI.

Such a large performance gap relative to JSC on this operation is worth digging into though. It's possible that JSC has special fast paths for some of those operations that give it an edge even in interpreted mode.

**Comment by mrousavy** [comment]
@neildhar after a while of profiling and debugging, I figured out what's so slow. It's the pure JS implementation of `pbkdf2`.

The one from `crypto`/`ethers` (polyfilled, pure JS) takes **14 seconds**:

```
ether's pbkdf2 took 14279.91887497902ms
```

and my pure C++ JSI implementation takes **0.001 second**:

```
C++ pbkdf2 took 1.4742500185966492ms
```

..so it's almost **10.000x faster**.

Still, Hermes is twice as slow as JSC, do you want me to leave that issue open?

**Comment by neildhar** [comment]
Thanks for the update and investigating further. Do you have a minimal repro you can share that invokes `pbkdf2` and shows the discrepancy between Hermes and JSC? 

I also wanted to check if you've also done a comparison on Android?

**Comment by mrousavy** [comment]
Yes, [here is the exact code snippet](https://github.com/ethers-io/ethers.js/blob/b1458989761c11bf626591706aa4ce98dae2d6a9/packages/pbkdf2/lib/browser-pbkdf2.js#L6-L45) that takes **14 seconds** on Hermes on my iPhone 8. It is twice as fast on JSC.

I have replaced this exact function with my C++ JSI one and went from **14 seconds** to **1 millisecond**.

I do not have tested Android yet, will do that soon.

**Comment by mrousavy** [comment]
I'll send you a list of example arguments you can use for that function in a sec.

**Comment by osdnk** [comment]
If you're interested in some non-ideal solution for your specific case, we used, in Rainbow, a native library we made for creating seeds, and then used `ethers.Wallet.fromMnemonic`.

**Comment by mrousavy** [comment]
@osdnk but `ethers.Wallet.fromMnemonic` is the slow part? On my iPhone 8 with Hermes it takes 14 seconds...

**Comment by osdnk** [comment]
`ethers.Wallet.fromMnemonic` with already pre-made mnemonics? It's interesting, for us this is s while, AFAIK

**Comment by mrousavy** [comment]
> `ethers.Wallet.fromMnemonic` with already pre-made mnemonics?

yes. 14 seconds. `ethers.Wallet.createRandom()` takes 33 seconds.

**Comment by osdnk** [comment]
so this is the time you had already with JSC?
